### PR TITLE
unstableGitUpdater: Use stableVersion, update to new format, offer hardcoded 0 version

### LINKED ISSUE #291729: unstableGitUpdater produces version numbers that do not comply with version naming guidelines
### Describe the bug
[the package naming rules](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-naming) say in no uncertain terms "The version attribute must start with a digit".

however, a very large number of packages do not follow this naming guideline.  almost all of these have a version starting with `unstable-`.  on one occasion i was even instructed to change a version starting with a digit to a version starting with `unstable-`

either common practice is wrong, or the documentation is wrong.

my suggested fix would be to tweak the documentation to match common practice: versions must begin with a number, or the string `unstable-` followed by a YYYY-MM-DD date.

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by OPNA2608**
## Description of changes

With #280312, we now have more complete guidelines on what unstable versions should look like in our post-#234201 world. As a first step towards that, update `unstableGitUpdater` update script - and all users of it, to avoid update errors - to actually use it.

- Remove `stableVersion` option, the stable version schema will always be used now.
  The only usage of this option inside of Nixpkgs that I could find (`gnomeExtensions.dash-to-dock`, back when it was manually packaged) was removed ~ 2 years ago. GitHub-wide code search was only able to find stale Nixpkgs clones with that one usage, so it *seems* fine to just drop it?
- Add a `hardcodeZeroVersion` option, for projects without any releases & incompatible tagging systems. This skips the tag search in favour of a hardcoded version `0`.

Not 100% sure yet if the way this is implemented in the script is the best way of doing it, but it's a start. Open for complaints / alternative solutions :slightly_smiling_face:.

Closes #291729

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

closes https://github.com/NixOS/nixpkgs/pull/294213

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by OPNA2608** [comment]
I have pushed a treewide change that updates all users of `unstableGitUpdater`:
- Added `hardcodeZeroVersion` or `tagPrefix` as needed, so the script can "find" a tag
- Updated the versions of affected packages to the new schema, so running updaters in bulk (for example `nix-shell maintainers/scripts/update.nix --argstr maintainer OPNA2608` doesn't error out on "the version is different, but the ref / hash is the same" situations

Running `maintainers/scripts/rebuild-amount.sh` says:
```
Estimating rebuild amount by counting changed Hydra jobs (parallel=unset).
    718 x86_64-darwin
   1029 x86_64-linux
```
…which *should* be fine for master, although more than I would've liked :sweat:.

I'll leave notes on individual packages where I think it's worth to point unusual / problematic things out.

File: pkgs/applications/audio/mmlgui/default.nix
Comment (OPNA2608) [inline pkgs/applications/audio/mmlgui/default.nix:87]: Tags started out fine with a nice `v1.0`, then they turned into a less-nice `YYMMDD-preview` before stopping altogether.

I would preferably have this be `1.0-unstable-YYYY-MM-DD`, but that would require teaching the updater about tags (patterns?) it should ignore, or telling it outright what version it should use. The former would need more effort to implement that and has the potential of patterns going stale without attentive maintainers, but might be worthwhile. The latter idea in comparison seems like a bad idea - more things the maintainers need to pay attention to, doesn't really follow the spirit of the guideline IMO.

File: pkgs/applications/office/treesheets/default.nix
Comment (OPNA2608) [inline pkgs/applications/office/treesheets/default.nix:15]: Copying from the commit description:

  For all intents and purposes, that seems to be their versioning now.

  - Last semver release 1.0.2 was more than 4 years ago
  - That's the version shown in the About dialogue
  - They previously said that they »do "releases" super infrequently or not at all«
  - Basically every commit is tagged like this, numbers are always-increasing

Due to that, I see no reason to use the `unstableGitUpdater` on this.

File: pkgs/applications/radio/chirp/default.nix
Comment (OPNA2608) [inline pkgs/applications/radio/chirp/default.nix:41]: This *could* be an argument for some logic in `unstableGitUpdater` itself to convert the tag based on a requested *thing* (pattern, code, whatever) into a regular `major[.minor[.patch]][whatever]`, but that is likely overkill (or at least, way more than what my energy levels permit me to do).

Maybe something like what `goperf` does where it first runs `unstableGitUpdater`, followed by some more code to update further things? The regular updater could write `x_y_z-unstable-YYY-MM-DD`, and the extra code could fix it up into `x.y.z-unstable-YYYY-MM-DD`?

Also, the `py3` branch seems to have been deleted so this still doesn't work, but I'm not sure what it should point at.

File: pkgs/by-name/ni/nix-search-cli/package.nix
Comment (OPNA2608) [inline pkgs/by-name/ni/nix-search-cli/package.nix:22]: Similar to `treesheets`, but IMO different because it more or less just expresses the same thing as `unstable-YYYY-MM-DD`. `treesheets` actually uses its weird number internally and presents it to the user in the About dialogue, `nix-search-cli` doesn't seem to use this data.

No strong opinion on this though, I'd leave that up to the maintainer / maybe upstream themselves to decide.

File: pkgs/development/compilers/open-watcom/v2.nix
Comment (OPNA2608) [inline pkgs/development/compilers/open-watcom/v2.nix:92]: Similar to `nix-search-cli` in terms of numeric tags, but the 3 latest ones are always numberless tags that get continuously recycled by CI. No real information one can extract from that system, so I think `0` is perfectly adequate here.

File: pkgs/development/libraries/cegui/default.nix
Comment (OPNA2608) [inline pkgs/development/libraries/cegui/default.nix:50]: Same as `chirp`: Might need post-updater code to convert `x-y-z-unstable-YYYY-MM-DD` into `x.y.z-unstable-YYYY-MM-DD`.

File: pkgs/development/tools/goperf/default.nix
Comment (OPNA2608) [inline pkgs/development/tools/goperf/default.nix:29]: Something must've changed about this at some point, because the custom updater code here was already broken before I touched anything:

```
error: attribute 'go-modules' in selection path 'goperf.go-modules.drvAttrs.outputHash' not found
       Did you mean goModules?
```

Otherwise an interesting approach that could be modified to handle weirder tag formats on a per-package basis.

File: pkgs/development/tools/misc/c2ffi/default.nix
Comment (OPNA2608) [inline pkgs/development/tools/misc/c2ffi/default.nix:15]: This ends up with 2 LLVM 11 version numbers in its full name, due to the previous tag already being a base LLVM version. The one in `pname` can likely be dropped.

`c2ffi-llvm-11.0.0-11.0.0.0-unstable-2021-06-15`

File: pkgs/development/tools/misc/luarocks/luarocks-nix.nix
Comment (OPNA2608) [inline pkgs/development/tools/misc/luarocks/luarocks-nix.nix:5]: The tags inherited from upstream `luarocks` only go up to `3.7.0`, but the master branch is clearly based on 3.9.2:

image

…so there's no winning here, everything is wrong. :slightly_frowning_face:

File: pkgs/by-name/sa/samrewritten/package.nix
Comment (OPNA2608) [inline pkgs/by-name/sa/samrewritten/package.nix:14]: This project uses `YYYYMM` tags, and puts `x.y` versions into its release descriptions… kind of upsetting :confused:.

But [nobody on repology uses the `x.y` versions](https://repology.org/project/samrewritten/versions), and Gentoo uses the `YYYYMM` tag number, so I guess we'll go with that.

**Review by OPNA2608 — COMMENTED** [review]
- Dates may have changed by a day here and there. That's based on the update script complaining about the day being off, which made the update process error out on "different version, same rev / hash".
- For projects on sourcehut I had to make guesses, due to the outage they're currently dealing with.

**Comment by kirillrdy** [comment]
@OPNA2608 greetings !

first, thank you for working on this. you have clearly put fair bit of work into this! Good stuff

initial impression is, that this approach is at least not very intrusive ( eg easy to change )

my first question would be, can we avoid `hardcodeZeroVersion = true` for repositories that clearly don't have any tags ?

**Comment by OPNA2608** [comment]
Because we default to using a shallow clone (`shallowClone ? true`), most of the time we start with an incomplete/missing list of all tags. The update script handles this by incrementally fetching more commits & associated tags:
```sh
depth=100
while (( $depth < 10000 )); do
  # <check & break if tag found>

  ${git}/bin/git fetch --depth="$depth" --tags
  depth=$(( $depth * 2 ))
done
```

The only 100% accurate ways of determining if the repo has *any* tags that I can see is if we either:
- don't use a shallow clone so we have all tags, at the cost of always requiring more time & space to fetch, or
- fetch everything once the 10000 incremental fetching range has been reached, going "we tried the quick way but it failed, now let's go the thorough route"

(Otherwise I'm not sure if the updater wouldn't just break / produce an incorrect bump if a repo had a 1.2.3 tag 10001 commits ago. I would assume it would happily submit `foobar: 1.2.3-unstable-2024-02-05 -> 0-unstable-2024-02-06`.)

I can add that, so we can avoid `hardcodeZeroVersion = true` for tag-less repos.

**Comment by kirillrdy** [comment]
- this is scoped to unstableGitUpdater
- there are significantly more derivations(projects) with no tags, should we assume `hardcodeZeroVersion` as default.

as for repos with some tag
- maintainers of derivations opt in to fetch tags
- Or we pass version to updater, and updater will prepend that version
- if a derivation is using `unstableGitupdater`, maybe version is no longer relevant, and `0` is ok to use

@OPNA2608 what do you think ?

**Comment by OPNA2608** [comment]
>     * there are significantly more derivations(projects) with no tags, should we assume `hardcodeZeroVersion` as default.
>
>
> as for repos with some tag
> 
>     * maintainers of derivations opt in to fetch tags
> 
>     * Or we pass version to updater, and updater will prepend that version
> 
>     * if a derivation is using `unstableGitupdater`, maybe version is no longer relevant, and `0` is ok to use

I think most of these aren't really bringing us closer to the goal of the guideline.

- The point is to have a real base version. Defaulting to lying about it by hardcoding 0 and requiring maintainers to opt into "real" versions seems to me like a bad design.
- Just because there are currently no tags doesn't mean that there will *never* be tags. I think having the version in a bump jump from zero to non-zero would be a good indicator that the derivation could potentially stop using the unstable schema.
- Allowing a hardcoded base version to be passed feels like something that will be set once and never updated. Hardcoding 0 despite tags, I hope, would only be used for wholly incompatible tagging systems.

When there are no tags at all, then i think adding the fallback is the best option - that saves us a `hardcodeZeroVersion` for the most common situation.

File: pkgs/development/tools/goperf/default.nix
Comment (kirillrdy) [reply pkgs/development/tools/goperf/default.nix:26]: `goperf.goModules` is correct

**Comment by kirillrdy** [comment]
@OPNA2608 what do you think about https://github.com/NixOS/nixpkgs/issues/291729 ?

**Comment by eclairevoyant** [comment]
> @eclairevoyant what are your thoughts on this PR

I haven't had a chance to look in depth; it looks involved and I like to avoid reviewing when I'm low on sleep :)

**Comment by OPNA2608** [comment]
I have added 2 more options, which should help us with covering possible situations:

- `tagFormat`: Filter tags to match a certain `grep -Eo` format. This should help with upstreams that provide pointless `latest` tags.
  - I don't have an immediate use for this inside of Nixpkgs, but I have seen upstreams that do this.
- `tagConverter`: Points to a command that should convert a prefix-stripped tag passed via stdin into our version format on stdout. This should provide better flexibility in handling tagging formats like `release_x_y_z`.
  - We have a single user of this, `chirp`.
  - `cegui` *would* also use this, if we weren't using a branch that lacks any tags. It's kind of pointlessly set there, as a "this is what we *would* do" kinda thing.

With that, the only packages that need to rely on `hardcodeZeroVersion` and their reasons are:

- `mpvScripts.mpv-webm`: Upstream only provides a constantly-retagged `latest` tag
- `open-watcom-v2-unwrapped`: Upstream provides monthly `YYYY-MM-DD-Build`, and constantly-retagged `Current-build`, `Last-CI-build` & `Coverity-scan` tags
  - Potentially a use case for `tagFormat` if `gitUpdater` supported it
- `nelua`: Upstream provides a very stale `latest` tag, and a single random `YYYYMMDD` tag from a month ago. Not sure what upstream's strategy is, and there was no compatible tag for the longest time, so keeping it at `0` to be safe.
- `c2ffi`: Upstream provided tags matching their LLVM base version, but stopped doing so 2 years ago. The full name for the current version would be `c2ffi-llvm-16.0.0-12.0.0.0-unstable-2023-11-18`… which is pure nonsense.
- `lua{51,52,53,54,jit}Packages.luarocks-nix`: The base `luarocks` provides consistent tags. `nix-community/luarocks-nix` only merges their upstream periodically and doesn't inherit the tags anymore, so the tags on that repo are lying. So I felt it made the most sense to not engage with the versions & tags and hardcode `0`. For reference:
  - Latest inherited tag from upstream luarocks: `v3.7.0`, 2021-04-13
  - Latest custom tag: `nix_v3.5.0-1`, 2021-07-28
  - Latest luarocks version [actually merged into the tree](https://github.com/nix-community/luarocks-nix/commit/a1bcd18dd3a61b67d595224536ab0cfe6bd59287): `v3.9.2`, 2022-12-08
- `nix-search-cli`: Almost every commit is tagged as `release-<unix timestamp>-<commit>`, software doesn't seem to keep track of its version in any way. Hardcoded `0` matches the upstream intention the closest IMO.

File: pkgs/common-updater/unstable-updater.nix
Comment (eclairevoyant) [inline pkgs/common-updater/unstable-updater.nix:112]: shouldn't this also respect `tag_format`?

File: pkgs/common-updater/unstable-updater.nix
Comment (eclairevoyant) [inline pkgs/common-updater/unstable-updater.nix:115]: and this as well?

File: pkgs/common-updater/unstable-updater.nix
Comment (eclairevoyant) [inline pkgs/common-updater/unstable-updater.nix:111]: `git` isn't on PATH, is it? would this work?

Maybe we should convert this to `writeShellApplication`, so we can use `runtimeInputs` and not have to repeat `${git}/bin/git` or `${lib.getExe git}` everywhere - and get the `shellcheck` benefits of that builder.

File: pkgs/common-updater/unstable-updater.nix
Comment (eclairevoyant) [inline pkgs/common-updater/unstable-updater.nix:127]: I find it a bit unintuitive that if both `tag_prefix` and `tag_converter` are provided, that we run them serially.

I think we should have an assert that at most one of these are used.

**Review by eclairevoyant — COMMENTED** [review]
I still haven't tested this, but just reading over the script, I found some points of feedback:

File: pkgs/common-updater/unstable-updater.nix
Comment (eclairevoyant) [reply pkgs/common-updater/unstable-updater.nix:112]: also, we can simplify the `tag_format` check by passing `--match` to the git command

Comment (OPNA2608) [reply pkgs/common-updater/unstable-updater.nix]: Implemented with `--match` now, and uses `--match` here.

File: pkgs/common-updater/unstable-updater.nix
Comment (OPNA2608) [reply pkgs/common-updater/unstable-updater.nix]: Uses `--match` here now.

File: pkgs/common-updater/unstable-updater.nix
Comment (OPNA2608) [reply pkgs/common-updater/unstable-updater.nix]: Refactored it to use `writeShellApplication`, added used packages to `runtimeInputs` and fixed as much as I could about the shellcheck messages.

File: pkgs/common-updater/unstable-updater.nix
Comment (OPNA2608) [reply pkgs/common-updater/unstable-updater.nix]: Added an assert so `tagPrefix` and `tagConverter` may not be passed together, and removed the usage of `tag_prefix` in the `tag_converter` case.

**Comment by lolbinarycat** [comment]
once this gets merged the new options will need to be documented in the nixpkgs manual, although for the sake of getting this done, my vote is that can be done in a seperate PR

**Comment by OPNA2608** [comment]
> once this gets merged the new options will need to be documented in the nixpkgs manual

I don't think `unstableGitUpdater` is documented *at all* in the (Nixpkgs) manual? The most I can find about our updaters in general is:

> A common pattern is to use the [`nix-update-script`](https://github.com/NixOS/nixpkgs/blob/master/pkgs/common-updater/nix-update.nix) attribute provided in Nixpkgs, which runs [`nix-update`](https://github.com/Mic92/nix-update):
> 
> ```
> passthru.updateScript = nix-update-script { };
> ```

**Comment by lolbinarycat** [comment]
well, then someone (maybe i'll give it a shot) should probably write a section about updater scripts in nixpkgs.

however, like i said, this should be doable in a separate PR after this one, and i think the fact it already isn't documented only strengthens that.

File: pkgs/development/libraries/cegui/default.nix
Comment (lolbinarycat) [inline pkgs/development/libraries/cegui/default.nix:56]: this seems like a shame, but i don't know how to fix it.

**Review by lolbinarycat — APPROVED** [review]
all 109 files checked, seems pretty solid.

File: pkgs/common-updater/unstable-updater.nix
Comment (lolbinarycat) [inline pkgs/common-updater/unstable-updater.nix:121]: the one thing that feels a bit sloppy is the fact that this long `git describe` command is repeated several times, but i don't think that's bad enough to block a PR over.

**Comment by lolbinarycat** [comment]
perhaps if you factored the changes to `unstableGitUpdater` out into a seperate PR (without the changes to version numbers) this would be easier to merge?  that should also bring the number of rebuilds down from thousands to zero.

File: pkgs/common-updater/unstable-updater.nix
Comment (OPNA2608) [reply pkgs/common-updater/unstable-updater.nix]: Moved into one bash function, that gets called where needed.

**Comment by OPNA2608** [comment]
As much as I dislike temporarily breaking updaters by not fixing all uses of `unstableGitUpdater` together with this, you're prolly right… Rebuild counts aside, some of the packages get auto-updated very frequently which causes merge conflicts, and I kinda don't have as much spare time as I'd like to keep this updated every day.

Leaving only the changes to `unstableGitUpdater` here, will submit the treewide changes to everything afterwards.

**Comment by eclairevoyant** [comment]
Oh! Did not realise this was as yet unmerged. Apologies.
Thanks for the work on this!

**Pull request merged**
